velox/exec/Window.cpp

```diff
@@ -379,25 +450,71 @@ void Window::callApplyForPartitionRows(
         }
         break;
       }
-      case core::WindowNode::BoundType::kPreceding:
-      case core::WindowNode::BoundType::kFollowing:
-        VELOX_NYI("Not supported");
+      case core::WindowNode::BoundType::kPreceding: {
+        kBoundFrameCheck(frameArg);
+        if (frameArg->isConstantValue) {
+          std::iota(
+              rawFrameBounds,
+              rawFrameBounds + numRows,
+              startRow - frameArg->constantValue - firstPartitionRow);
+        } else {
+          updateFrameVector(frameArg.value());
+          for (auto i = 0; i < numRows; i++) {
+            auto offset = frameArg->columnarOffset->valueAt(i);
+            rawFrameBounds[i] = startRow - offset - firstPartitionRow;
+          }
+        }
+        break;
+      }
+      case core::WindowNode::BoundType::kFollowing: {
+        kBoundFrameCheck(frameArg);
+        if (frameArg->isConstantValue) {
+          std::iota(
+              rawFrameBounds,
+              rawFrameBounds + numRows,
+              startRow + frameArg->constantValue - firstPartitionRow);
+        } else {
+          updateFrameVector(frameArg.value());
+          for (auto i = 0; i < numRows; i++) {
+            auto offset = frameArg->columnarOffset->valueAt(i);
+            rawFrameBounds[i] = startRow + offset - firstPartitionRow;
+          }
+        }
+        break;
+      }
       default:
         VELOX_USER_FAIL("Invalid frame bound type");
     }
   };
 
+  auto emptyFrameCheck = [&](vector_size_t* frameStart,
```

Comment:
Call this function fixEmptyFrames. Please also explain why this code is needed. Can we possibly do this check only for a subset of the frames ? If yes, then maybe we should do it only conditionally.